main.vala

```diff
@@ -639,291 +514,1244 @@ class Vls.Server {
+            }
+        }
+    }
+
+    Vala.CodeNode get_best (FindSymbol fs, TextDocument file) {
+        Vala.CodeNode? best = null;
+
+        foreach (var node in fs.result) {
+            if (best == null) {
+                best = node;
+            } else {
+                var best_begin = new Position.from_libvala (best.source_reference.begin);
+                var best_end = new Position.from_libvala (best.source_reference.end);
+                var node_begin = new Position.from_libvala (node.source_reference.begin);
+                var node_end = new Position.from_libvala (node.source_reference.end);
+
+                if (best_begin.compare (node_begin) <= 0 && node_end.compare (best_end) <= 0 &&
+                    // don't get implicit `this` accesses
+                    !(best.source_reference.begin.column == node.source_reference.begin.column &&
+                        node.source_reference.end.column == best.source_reference.end.column &&
+                        node is Vala.MemberAccess && 
+                        ((Vala.MemberAccess)node).member_name == "this" &&
+                        ((Vala.MemberAccess)node).inner == null))
+                    best = node;
             }
+        }
+
+        assert (best != null);
+        var sr = best.source_reference;
+        var from = (long)Server.get_string_pos (file.content, sr.begin.line-1, sr.begin.column-1);
+        var to = (long)Server.get_string_pos (file.content, sr.end.line-1, sr.end.column);
+        string contents = file.content [from:to];
+        debug ("Got best node: %s @ %s = %s", best.type_name, sr.to_string(), contents);
+
+        return (!) best;
+    }
+
+    Vala.Scope get_current_scope (Vala.CodeNode code_node) {
+        Vala.Scope? best = null;
 
-            debug (@"textDocument/publishDiagnostics: $uri");
+        for (Vala.CodeNode? node = code_node; node != null; node = node.parent_node) {
+            if (node is Vala.Symbol) {
+                var sym = (Vala.Symbol) node;
+                best = sym.scope;
+                break;
+            }
         }
+
+        assert (best != null);
+
+        return (!) best;
     }
 
     void textDocumentDefinition (Jsonrpc.Server self, Jsonrpc.Client client, string method, Variant id, Variant @params) {
         var p = parse_variant <LanguageServer.TextDocumentPositionParams> (@params);
-        debug ("get definition in %s at %u,%u", p.textDocument.uri,
-            p.position.line, p.position.character);
-        var sourcefile = ctx.get_source_file (p.textDocument.uri);
+        TextDocument? sourcefile = lookup_source_file (p.textDocument.uri);
         if (sourcefile == null) {
-            debug ("unknown file %s", p.textDocument.uri);
+            debug (@"[$method] file `$(p.textDocument.uri)' not found");
+            reply_null (id, client, method);
+            return;
+        }
+        var file = sourcefile.file;
+
+        wait_for_context_update (id, request_cancelled => {
```

Comment:
why do you wait_for_context_update here?

Reply:
Because we can't show definitions until the code context's symbol tree matches the source code. We update the source code immediately every time we get `textDocument/didChange`, but we refresh the code context on an interval.